### PR TITLE
New UI for Metamist

### PR DESCRIPTION
**Pull request by danielreti**
Draft for upcoming UI changes 

Completed Changes
- [x] Added Search to Nav Bar
- [x] Redesigned Nav Bar
- [x] Added links to project overview page
- [x] Created Sample page with info + pedigree
- [x] Included Project-level statistics
- [x] Fixed routing + implemented hotlink URLs
- [x] Add new interfaces to API

Upcoming
- [ ] Fix linting
- [ ] Add participant and family pages
- [ ] Improve pedigrees
- [ ] Add CRAM/file/seqr info (requires more data + endpoints)

### CODE REVIEW

File: web/src/Search.tsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    const handleSearchChange = React.useCallback((e, data) => {
+        clearTimeout(timeoutRef.current as NodeJS.Timeout);
+        dispatch({ type: ActionKind.Start, query: data.value });
+
+        timeoutRef.current = setTimeout(async () => {
+            if (data.value.length === 0) {
+                dispatch({ type: ActionKind.Clear });
+                return;
+            }
+            new WebApi().searchByKeyword(data.value).then((resp) => {
```
Comment (illusional) [inline web/src/Search.tsx:79]: Need to add a `catch` to this

File: web/src/Search.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    loading: boolean;
+    results: {}[];
+    value: string;
+};
+
+enum ActionKind {
+    Clear = "CLEAN_QUERY",
+    Start = "START_SEARCH",
+    Finish = "FINISH_SEARCH",
+    Update = "UPDATE_SELECTION",
```
Comment (illusional) [inline web/src/Search.tsx:16]: Handle error case?

File: web/src/Search.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            input={{ fluid: true }}
+            loading={loading}
+            placeholder="Search..."
+            onResultSelect={(e, data) => {
+                dispatch({
+                    type: ActionKind.Update,
+                    selection: data.result.title,
+                });
+                navigate(
+                    `/project/${data.result.data.project}/${data.result.type}/${data.result.title}`,
```
Comment (illusional) [inline web/src/Search.tsx:108]: Could we add a `searchResultToRoute` function? Make a switch handle it to catch unimplemented cases and just give a bit more flexibility to route handling.

```suggestion
                    `/project/${data.result.data.project}/${data.result.type}/${data.result.data.id}`,
```

File: web/src/mystyle.css [hunk 4]
```diff
@@ -0,0 +1,54 @@
+.react-grid-layout {
```
Comment (illusional) [inline web/src/mystyle.css:1]: Could we give this file a better name?

File: web/src/project/ProjectSummary.tsx [hunk 5]
```diff
@@ -1,183 +1,364 @@
...
+    );
 
-    if (error) {
-        table = <p><em>An error occurred when fetching samples: {error}</em></p>
-    }
-    else if (!summary) {
-        if (globalContext.project) { table = <p>Loading...</p> }
-        else { table = <p><em>Please select a project</em></p> }
-    }
+    const renderLink = (s: any, k: any, external_id: string) => (
```
Comment (illusional) [inline web/src/project/ProjectSummary.tsx:105]: In previous projects I've added something like this to standardise SampleLinks, and then only have to change one place to change all the sample URLs. Could we add a collection of link to some standard place to make URL-ing standard?

```
export const SampleLink = ({id, projectName}) => <Link to `/project/{projectName}/sample/${id}`
```

File: web/src/project/ProjectSummary.tsx [hunk 6]
```diff
@@ -1,183 +1,364 @@
...
+        <Link to={`/project/${projectName}/sample/${external_id}`}>
```
Comment (illusional) [inline web/src/project/ProjectSummary.tsx:106]: This should be the internal sample ID (`CPGxxx`)

File: web/src/urls.tsx [hunk 7]
```diff
@@ -1,55 +1,56 @@
...
+                    element={<SwaggerUI url="/openapi.json" />}
+                />
+
+                <Route path="project/" element={<ProjectSummary />} />
+                <Route
+                    path="project/:projectName"
+                    element={<ProjectSummary />}
+                />
+                <Route
+                    path="project/:projectName/:page/:pageSize"
```
Comment (illusional) [inline web/src/urls.tsx:30]: IMO, pageSize should remain a query param

File: web/src/urls.tsx [hunk 8]
```diff
@@ -1,55 +1,56 @@
...
+                    path="project/:projectName/sample/:sampleName"
+                    element={<DetailedInfoPage />}
+                />
+                <Route
+                    path="project/:projectName/family/:familyName"
+                    element={<DetailedInfoPage />}
+                />
+                <Route
+                    path="project/:projectName/participant/:participantName"
+                    element={<DetailedInfoPage />}
```
Comment (illusional) [inline web/src/urls.tsx:51]: All of these should be the internal identifier for each, all of these will have multiple external IDs (in the future) which could potentially change.

File: web/src/project/ProjectSummary.tsx (hunk 5 above)
Comment (illusional) [reply web/src/project/ProjectSummary.tsx:105]:
```typscript
<SampleLink id="CPGLCL28">GRKMYTH11J1</SampleLink>
<SampleLink id="CPGLCL28" />
```

File: web/src/Search.tsx [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    const navigate = useNavigate();
+    const [{ loading, results, value }, dispatch] = React.useReducer(
+        exampleReducer,
+        initialState
+    );
+
+    const timeoutRef = React.useRef<NodeJS.Timeout | null>(null);
+
+    const handleSearchChange = React.useCallback((e, data) => {
+        clearTimeout(timeoutRef.current as NodeJS.Timeout);
```
[the remark this answers is not in the document]
Comment (danielreti) [reply web/src/Search.tsx:94]: Added

File: web/src/Search.tsx [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    results?: {}[];
+};
+
+const initialState: State = {
+    loading: false,
+    results: [],
+    value: "",
+};
+
+const exampleReducer = (state: State, action: Action): State => {
```
[the remark this answers is not in the document]
Comment (danielreti) [reply web/src/Search.tsx:32]: Done

File: web/src/Search.tsx [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        case ActionKind.Clear:
+            return initialState;
+        case ActionKind.Start:
+            return { ...state, loading: true, value: action.query! };
+        case ActionKind.Finish:
+            return { ...state, loading: false, results: action.results! };
+        case ActionKind.Update:
+            return { ...state, value: action.selection! };
+
+        default:
```
[the remark this answers is not in the document]
Comment (danielreti) [reply web/src/Search.tsx:43]: Replaced default with an unreachable function to catch any missing cases - is this what you meant?
```
const assertUnreachable = (x: ActionKind): never => {
    throw new Error(`Unexpected value: ${x}`);
};

    switch (action.type) {
        case "CLEAN_QUERY":
            return initialState;
        case "START_SEARCH":
            return { ...state, loading: true, value: action.query! };
        case "FINISH_SEARCH":
            return { ...state, loading: false, results: action.results! };
        case "UPDATE_SELECTION":
            return { ...state, value: action.selection! };
        case "QUERY_TOO_SHORT":
            return { ...state, loading: false };
    }
    return assertUnreachable(action.type);
```

File: web/src/Search.tsx (hunk 1 above)
Comment (danielreti) [reply web/src/Search.tsx:102]: Added a console.log  - how do you want these errors handled?
Also what are some examples of these errors?

File: web/src/Search.tsx (hunk 2 above)
Comment (danielreti) [reply web/src/Search.tsx:16]: TODO once I understand better

File: web/src/project/ProjectSummary.tsx (hunk 6 above)
Comment (danielreti) [reply web/src/project/ProjectSummary.tsx:107]: Fixed

File: web/src/urls.tsx (hunk 8 above)
Comment (danielreti) [reply web/src/urls.tsx:51]: Noted and fixed for sample page - other pages to do. But they don't have internal and external IDs right? (do they?)

File: web/src/Search.tsx (hunk 1 above)
Comment (danielreti) [reply web/src/Search.tsx]: Added a catch and an error dropdown
Trigger by adding a raise ValueError in the python search

File: web/src/Search.tsx (hunk 2 above)
Comment (danielreti) [reply web/src/Search.tsx:16]: As above, added an error case

File: web/package.json [hunk 12]
```diff
@@ -14,13 +14,16 @@
...
     "react-dom": "^17.0.2",
+    "react-grid-layout": "^1.3.4",
     "react-markdown": "^7.1.0",
     "react-router-dom": "^6.0.1",
     "react-scripts": "^4.0.3",
     "react-syntax-highlighter": "^15.4.4",
     "reactstrap": "^9.0.1",
     "remark-gfm": "^3.0.1",
     "remark-toc": "^8.0.1",
+    "semantic-ui-css": "^2.5.0",
```
[the remark this answers is not in the document]
Comment (danielreti) [reply web/package.json:29]: Switched all to semantic and removed `reactstrap`

File: web/src/Search.tsx (hunk 3 above)
Comment (danielreti) [reply web/src/Search.tsx]: Added a switch with alerts to handle unimplemented cases

File: web/src/project/ProjectSummary.tsx (hunk 5 above)
Comment (danielreti) [reply web/src/project/ProjectSummary.tsx]: Great suggestion - added

File: web/src/urls.tsx (hunk 7 above)
Comment (danielreti) [reply web/src/urls.tsx]: Fixed

File: web/src/DetailedInfoPage.tsx [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+    const [sampleInfo, setSampleInfo] = React.useState<Sample>();
+    const [pedigree, setPedigree] = React.useState<PedigreeEntry[]>();
+    const [participants, setParticipants] =
+        React.useState<ParticipantModel[]>();
+    const [selectedExternalID, setSelectedExternalID] =
+        React.useState<string>();
+    const [idNameMap, setIdNameMap] = React.useState<[string, string][]>();
+    const [sequenceInfo, setSequenceInfo] = React.useState<SampleSequencing>();
+
+    const getSamplesFromProject = React.useCallback(async () => {
```
Comment (illusional) [inline web/src/DetailedInfoPage.tsx:52]: I think this is fine for first iteration, but we should move all this logic to the web layer, and have a "get family display info for web" sort of thing.

**Review by illusional — APPROVED** [review]

File: web/src/DetailedInfoPage.tsx (hunk 13 above)
Comment (danielreti) [reply web/src/DetailedInfoPage.tsx:52]: Yeah definitely, would be great to have fewer (or 1) API call which gets the all relevant info

**Pull request merged**